### PR TITLE
Develop upstream sync 190311

### PR DESCRIPTION
**Pull request by sumesh13**
Upstream merge

### CODE REVIEW

**Comment by whchung** [comment]
@sumesh13 please rebase and fix errors. now dynamic loading branch #318 has been merged I expect previous changes for dynamic linking be removed.

**Comment by sumesh13** [comment]
Jenkins retest this please

**Comment by sumesh13** [comment]
Jenkins retest this please

**Comment by whchung** [comment]
@sumesh13 the PR looks to be in a pretty decent shape now. just a couple of failed unit tests to be figured out and we shall be good to take it.

**Comment by sumesh13** [comment]
Non XLa looks good. The fails are new tests added. However on the XLA side, the build failed for a big category of tests. Turns out there are some new cuda dependencies that have been added. Am working on fixing those instances. Need to iterate on rerun-fix....

File: tensorflow/compiler/xla/service/BUILD [hunk 1]
```diff
@@ -813,11 +813,15 @@ cc_library(
     deps = [
         ":service",
         "//tensorflow/compiler/xla/service/gpu:gpu_transfer_manager",
-        "//tensorflow/compiler/xla/service/gpu:nvptx_compiler",
         "//tensorflow/core:stream_executor_no_cuda",
+        "//tensorflow/compiler/xla/service/gpu:gpu_compiler",
```
Comment (whchung) [inline tensorflow/compiler/xla/service/BUILD]: not in this PR, but we probably need to break `gpu_compiler` into `nvptx_compiler` and `amdgpu_compiler`.

File: tensorflow/compiler/xla/service/gpu/BUILD [hunk 2]
```diff
@@ -571,44 +575,44 @@ tf_cc_test(
...
-        "//tensorflow/core:lib",
-        "//tensorflow/core:stream_executor_no_cuda",
-        "//tensorflow/stream_executor:blas",
-    ]
-    + if_cuda_is_configured(if_cuda([
-        "@local_config_cuda//cuda:cuda_headers",
-        "@local_config_cuda//cuda:cusolver",
-    ])),
-)
+#cc_library(
```
Comment (whchung) [inline tensorflow/compiler/xla/service/gpu/BUILD]: do we need to comment out this target? perhaps use `if_cuda_is_configured()` to make `srcs` and `hdrs` empty?

File: tensorflow/compiler/xla/service/gpu/BUILD [hunk 3]
```diff
@@ -571,44 +575,44 @@ tf_cc_test(
...
-        "//tensorflow/compiler/xla/service:hlo",
-        "//tensorflow/compiler/xla/service:hlo_pass",
-        "//tensorflow/core:lib",
-        "//tensorflow/core:stream_executor_no_cuda",
-        "//tensorflow/stream_executor:blas",
-        "@com_google_absl//absl/types:optional",
-    ],
-)
+#cc_library(
+#    name = "cusolver_rewriter",
```
Comment (whchung) [inline tensorflow/compiler/xla/service/gpu/BUILD]: do we need to comment out this target? perhaps use `if_cuda_is_configured()` to make `srcs` and `hdrs` empty?

File: tensorflow/compiler/xla/service/gpu/BUILD [hunk 4]
```diff
@@ -834,22 +838,25 @@ cc_library(
...
     alwayslink = True,  # Contains per-platform transfer manager registration
 )
 
 cc_library(
     name = "gpu_compiler",
     srcs = if_cuda_is_configured(["nvptx_compiler.cc"]) 
            + if_rocm_is_configured(["amdgpu_compiler.cc"]),
-    hdrs = if_cuda_is_configured(["nvptx_compiler.h"]) 
+    hdrs = [
+             "cudnn_batchnorm_rewriter.h",
```
Comment (whchung) [inline tensorflow/compiler/xla/service/gpu/BUILD]: @sumesh13 could you explain rationale behind changes wrt `cudnn_batchnorm_rewriter`. doesn't seem quite right to me.

File: tensorflow/compiler/xla/service/gpu/gpu_executable.cc [hunk 5]
```diff
@@ -48,18 +48,18 @@ using tensorflow::tracing::ScopedAnnotation;
 // Implementation note: HLO profiling is always enabled for GPU executables,
 // since we can use timers around thunks.
 GpuExecutable::GpuExecutable(
-    const string& ptx, const std::vector<uint8>& cubin,
-    std::pair<int, int> compute_capability,
+    const string& text, const std::vector<uint8>& binary,
+//    std::pair<int, int> compute_capability,
```
Comment (whchung) [inline tensorflow/compiler/xla/service/gpu/gpu_executable.cc]: @sumesh13 could you explain why we need to remove `compute_capability` here. I dislike this field very much as well but do we have to remove it?

File: tensorflow/compiler/xla/service/gpu/gpu_executable.cc [hunk 6]
```diff
@@ -70,6 +70,7 @@ Status GpuExecutable::ExecuteThunks(
   se::Stream* main_stream = run_options->stream();
   se::StreamExecutor* executor = main_stream->parent();
 
+#if 0
```
Comment (whchung) [inline tensorflow/compiler/xla/service/gpu/gpu_executable.cc]: `#if !TENSORFLOW_USE_ROCM`?

File: tensorflow/compiler/xla/service/gpu/gpu_transfer_manager.cc [hunk 7]
```diff
@@ -178,16 +182,35 @@ Status GpuTransferManager::TransferLiteralFromOutfeed(
 }  // namespace gpu
 }  // namespace xla
 
+
+#if 0
```
Comment (whchung) [inline tensorflow/compiler/xla/service/gpu/gpu_transfer_manager.cc]: `#if !TENSORFLOW_USE_ROCM`?

File: tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc [hunk 8]
```diff
@@ -39,7 +39,7 @@ limitations under the License.
 #include "tensorflow/compiler/xla/service/buffer_assignment.h"
 #include "tensorflow/compiler/xla/service/dfs_hlo_visitor.h"
 #include "tensorflow/compiler/xla/service/gpu/buffer_allocations.h"
-#include "tensorflow/compiler/xla/service/gpu/cholesky_thunk.h"
+//#include "tensorflow/compiler/xla/service/gpu/cholesky_thunk.h"
```
Comment (whchung) [inline tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc]: `#if !TENSORFLOW_USE_ROCM`? or is it possible to still include the header?

File: tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc [hunk 9]
```diff
@@ -487,6 +487,7 @@ Status IrEmitterUnnested::HandleCustomCall(HloInstruction* custom_call) {
     return Status::OK();
   }
 
+#if 0
```
Comment (whchung) [inline tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc]: `#if !TENSORFLOW_USE_ROCM`?

File: tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc [hunk 10]
```diff
@@ -2158,20 +2160,17 @@ std::unique_ptr<Thunk> IrEmitterUnnested::BuildConditionalThunk(
...
-  TF_CHECK_OK(true_computation->Accept(&ir_emitter_true));
-
-  HloComputation* false_computation = hlo->false_computation();
-  IrEmitterUnnested ir_emitter_false(
-         hlo_module_config_, 
-         false_computation,
-         ir_emitter_context_,
-          GetTargetIRBuilder().GetTargetMachineFeatures());
-  TF_CHECK_OK(false_computation->Accept(&ir_emitter_false));
+  std::vector<BufferAllocation::Slice> branch_operands;
```
Comment (whchung) [inline tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc]: @sumesh13 I think the updated logic which stores into `branch_thunks` is the new logic, right?

File: tensorflow/compiler/xla/tests/llvm_compiler_test.cc [hunk 11]
```diff
@@ -18,7 +18,11 @@ limitations under the License.
 #include "tensorflow/compiler/xla/literal_util.h"
 #include "tensorflow/compiler/xla/service/backend.h"
 #include "tensorflow/compiler/xla/service/cpu/cpu_compiler.h"
-#include "tensorflow/compiler/xla/service/gpu/nvptx_compiler.h"
+#if GOOGLE_CUDA
```
Comment (whchung) [inline tensorflow/compiler/xla/tests/llvm_compiler_test.cc]: `#if !TENSORFLOW_USE_ROCM` . let's try avoid `GOOGLE_CUDA` within XLA.

File: tensorflow/compiler/xla/tests/llvm_compiler_test.cc [hunk 12]
```diff
@@ -138,7 +142,11 @@ class CpuCompilerTest : public LLVMCompilerTest {
 
 class GpuCompilerTest : public LLVMCompilerTest {
  public:
+#if GOOGLE_CUDA
```
Comment (whchung) [inline tensorflow/compiler/xla/tests/llvm_compiler_test.cc]: `#if !TENSORFLOW_USE_ROCM` . let's try avoid `GOOGLE_CUDA` within XLA, as there really is no CUDA within XLA.

File: tensorflow/compiler/xla/tests/llvm_compiler_test.cc [hunk 13]
```diff
@@ -147,7 +155,11 @@ TEST_F(CpuCompilerTest, HooksTest) {
 }
 
 TEST_F(GpuCompilerTest, HooksTest) {
+#if GOOGLE_CUDA
```
Comment (whchung) [inline tensorflow/compiler/xla/tests/llvm_compiler_test.cc]: #if !TENSORFLOW_USE_ROCM . let's try avoid GOOGLE_CUDA within XLA, as there really is no CUDA within XLA but only NVPTX.

File: tensorflow/compiler/xla/tests/llvm_compiler_test.cc [hunk 14]
```diff
@@ -156,8 +168,13 @@ TEST_F(CpuCompilerTest, CpuMultiModuleCompilation) {
   TestMultiModuleCompilation(&compiler);
 }
 
-TEST_F(GpuCompilerTest, NVPTXMultiModuleCompilation) {
+TEST_F(GpuCompilerTest, GpuMultiModuleCompilation) {
+#if GOOGLE_CUDA
```
Comment (whchung) [inline tensorflow/compiler/xla/tests/llvm_compiler_test.cc]: #if !TENSORFLOW_USE_ROCM . let's try avoid GOOGLE_CUDA within XLA, as there really is no CUDA within XLA but only NVPTX.

**Comment by sumesh13** [comment]
I am trying to get a initial set of results.  I do plan to go over some of these to clean up. There is more cuda dependence introduced from upstream.

**Comment by sumesh13** [comment]
Am left with two fails - one due to some changes around scratch_allocator. Will come back to clean up after that

**Comment by whchung** [comment]
@sumesh13 some minor conflicts to be addressed with upstream

**Comment by whchung** [comment]
ping?

**Comment by sumesh13** [comment]
Not converged yet ...some progress on the xla  fails but not enough for clean run yet...

**Comment by whchung** [comment]
@sumesh13  All 3 paths still fail. Please help look into it ASAP. The PR itself is already getting old.

If necessary, get non-XLA path working first and fix XLA path in a subsequent PR.

@deven-amd have submitted a PR to address failures on CUDA path. Please consider rebase with `develop-upstream` and adopt those fixes.

**Comment by whchung** [comment]
@sumesh13 

Reviewing commit history of this PR it seems the very first merge commit:

https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/pull/351/commits/ab84f11b7c0c056a9ae1dc0c43a873c2c1fc9f2e

had quite a few issues that quite a few ROCm-specific logic in `develop-upstream` branch were removed in that commit, thus in the later commits they have to be reinstated.

Please help consider if the PR, in its current form is salvageable. Or should we start a new round afresh.

**Comment by sumesh13** [comment]
I realized there were some logic missing. Was trying to fix that. That doesnt seem to be converging. So I have started a new one. Checking with Deven on a eigen issue that I am running into.

**Comment by sumesh13** [comment]
I opened https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/pull/358

**Pull request closed without merging**